proposed/resource-limit.md

```diff
@@ -0,0 +1,559 @@
+
+However, these limits may be more complex and require that the users to explicitly release the resources upon completion of their operation. Given the trade-off between consistency and speed, it is also likely to be poorly suited to limiting resource usage among different services or machines.
+
+### Unified abstraction
+
+We want to use a single abstraction to represent both types of limits. This is because for many applications, there's a desire to represent both limits in a transparent way. For example, in ASP.NET Core we want to allow users to register a set of limits before processing a specific request and only proceed if all limits are satisfied. These limits may be rate or concurrency based or even based on some ambient system state. In order to represent all of these scenarios in a way that can be easily queried, a unified API seems more appropriate, rather than two separate APIs for self replenishing resource and non self-replenishing resources and have applications track these resources independently.
+
+### Composition with BCL types
+
+There are two main reasons to introduce this API in the BCL.
+
+#### Shared abstractions and implementation for the entire stack
+
+By introducing the new API in the BCL, we maximize the ability to share resource limiter across BCL types in additional to usage in ASP.NET Core and user code. For example, rate limiters can be used to limit traffic in BCL types, ASP.NET Core middlewares and service implementations.
+
+#### Integration with BCL types for ease of usage
+
+There are various ways to compose usage of the proposed resource limiting APIs and BCL types. For example, there is a trade-off between coupling and ease of usage. Consider the options:
+
+```c#
+// No coupling
+var channel = Channel.CreateBounded<string>(5);
+var limiter = new SampleRateLimiter(resourceCount: 5, newResourcePerSecond: 5);
+
+if (limiter.TryAcquire(out _)) {
+    channel.Writer.TryWrite("New message");
+}
+
+// Coupling
+var limiter = new SampleRateLimiter(resourceCount: 5, newResourcePerSecond: 5);
+// This creates a LimitedChannel<T>. Its operations are limited by the limiter specified,
+// which may be rate/concurrency based. LimitedChannel<T> with a concurrency limiter would
+// be similar to a BoundedChannel if the limiter's queue is disabled, see design section
+// for details on limiter's queue.
+var channel = Channel.CreateLimited<string>(limiter);
+
+channel.Writer.TryWrite("New message");
+```
+
+The benefit of tighter coupling with the BCL types is that it allows for simpler usage. In the former case, the user need to explicitly check the limiter every time an operation to the channel writer is requested. In the latter case, the channel writer can be used in the same manner as today and the channel writer itself will perform the check on the resource limiter on behalf of the user.
```

Comment:
I do not think that this example demonstrates the need to the tight coupling.

Why this cannot be:
```
var channel = new ResourceLimitedChannel<string>(limitter);

channel.Writer.TryWrite("New message");
```
No coupling required. This is similar pattern used in many other places in BCL. One example from many - `Stream` vs. `BufferedStream`.

Reply:
Agreed, I don't think that's the point though. I think the point is that there are places we would want to be able to both use in the implementation and take as public API as part of the assemblies where these existing BCL types exist. So in your `ResourceLimitedChannel` example, that would live with the channels package and would reference these types. Same with pipelines

Reply:
I can believe you. The examples in the doc do not demonstrate the point well.

Reply:
@davidfowl - concretely, is the only concern whether it is `Channel.CreateLimited(...)` vs. `LimitedChannel.Create(...)`?

Either way, you can still pass around the base `Channel` types, whether the limited Channel implementation lives in the Channels assembly or not.

Reply:
Where it lives matters. When you have 2 types that don't want to take a dependency on each other you force everyone to what I call "bridge packages". Then you end up with a bridge package per concept if those concepts don't live at the same layer in the dependency graph. For example:

- HttpClient + ResourceLimits = one package
- Channels  + ResourceLimits = one package 
- Pipelines + ResourceLimits = one package

You see the pattern.